docs/datagen/server/recipes.md

```diff
@@ -0,0 +1,173 @@
+
+After implementation, the provider must be [added][datagen] to the `DataGenerator`.
+
+`RecipeBuilder`
+---------------
+
+`RecipeBuilder` is a convenience implementation for creating `FinishedRecipe`s to generate. It provides basic definitions for unlocking, grouping, saving, and getting the result of a recipe. This is done through `#unlockedBy`, `#group`, `#save`, and `#getResult` respectively.
+
+!!! important
+    [`ItemStack` outputs][stack] in recipes are not supported within vanilla recipe builders for `RecipeProvider`. A `FinishedRecipe` must be built in a different manner for existing vanilla recipe serializers to generate this data.
+
+!!! warning
+    The item results being generated must have a valid `CreativeModeTab` specified; otherwise, a `NullPointerException` will be thrown. If no category accurately represents the item and no tab should be created, set the `#tab` property during item initialization to `CreativeModeTab#TAB_SEARCH`.
+
+All recipe builders except for [`SpecialRecipeBuilder`] require an advancement criteria to be specified. All recipes generate a criteria unlocking the recipe if the player has used the recipe previously. However, an additional criteria must be specified that allows the player to obtain the recipe without any prior knowledge. If any of the criteria specified is true, then the played will obtain the recipe for the recipe book.
+
+!!! tip
+    Recipe criteria commonly use `InventoryChangeTrigger` to unlock their recipe when certain items are present in the user's inventory.
+
+### ShapedRecipeBuilder
+
+`ShapedRecipeBuilder` is used to generate shaped recipes. The builder can be initialized via `#shaped`. The recipe group, input symbol pattern, symbol definition of ingredients, and the recipe unlock criteria can be specified before saving.
+
+```java
+// In RecipeProvider#buildCraftingRecipes(writer)
+ShapedRecipeBuilder builder = ShapedRecipeBuilder.shaped(result)
+  .pattern("a a") // Create recipe pattern
+  .define('a', item) // Define what the symbol represents
+  .unlockedBy("criteria", criteria) // How the recipe is unlocked
+  .save(writer); // Add data to builder
+```
+
+#### Additional Validation Checks
+
+Shaped recipes have some additional validation checks performed before building:
+
+* A pattern must be defined and take in more than one item.
+* All pattern rows must be the same width.
+* A symbol cannot be defined more than once.
+* The space character (`' '`) is reserved for representing no item in a slot and, as such, cannot be defined.
+* A pattern must use all symbols defined by the user.
+* A criteria besides from using the recipe must be specified to unlock the recipe.
+
+### ShapelessRecipeBuilder
+
+`ShapelessRecipeBuilder` is used to generate shapeless recipes. The builder can be initialized via `#shapeless`. The recipe group, input ingredients, and how the recipe is unlocked can be specified before saving.
```

Comment:
Parallelism as per https://github.com/MinecraftForge/Documentation/pull/420#discussion_r845640222 (please also repeat this for the following sections):
```suggestion
`ShapelessRecipeBuilder` is used to generate shapeless recipes. The builder can be initialized via `#shapeless`. The recipe group, input ingredients, and the recipe unlock criteria can be specified before saving.
```